test/test_xnan_functions.cpp

```diff
@@ -261,109 +261,94 @@ namespace xt
         xarray<float> afloat(shape);
         xarray<double> adouble(shape);
 
-#define CHECK_RESULT_TYPE_FOR_ALL(INPUT, RESULT_TYPE)             \
-        CHECK_RESULT_TYPE(nansum(INPUT, {1, 2}), RESULT_TYPE);    \
-        CHECK_RESULT_TYPE(nanmean(INPUT, {1, 2}), double);        \
-        CHECK_RESULT_TYPE(nanvar(INPUT, {1, 2}), double);         \
-        CHECK_RESULT_TYPE(nanstd(INPUT, {1, 2}), double);
-
-#define CHECK_NAN_MIN_MAX_RESULT_TYPE(INPUT, RESULT_TYPE)         \
-        CHECK_RESULT_TYPE(nanmin(INPUT, {1, 2}), RESULT_TYPE)     \
-        CHECK_RESULT_TYPE(nanmax(INPUT, {1, 2}), RESULT_TYPE)
-
-#define CHECK_TEMPLATED_RESULT_TYPE_FOR_ALL(INPUT, TEMPLATE_TYPE, RESULT_TYPE, STD_TYPE)  \
-        CHECK_RESULT_TYPE(nansum<TEMPLATE_TYPE>(INPUT, {1, 2}), RESULT_TYPE)              \
-        CHECK_RESULT_TYPE(nanmean<TEMPLATE_TYPE>(INPUT, {1, 2}), RESULT_TYPE)             \
-        CHECK_RESULT_TYPE(nanvar<TEMPLATE_TYPE>(INPUT, {1, 2}), RESULT_TYPE)              \
-        CHECK_RESULT_TYPE(nanstd<TEMPLATE_TYPE>(INPUT, {1, 2}), STD_TYPE)
+#define CHECK_RESULT_TYPE_FOR_ALL(INPUT, RESULT_TYPE, MINMAX_TYPE)  \
+        CHECK_RESULT_TYPE(nansum(INPUT, {1, 2}), RESULT_TYPE);      \
+        CHECK_RESULT_TYPE(nanmean(INPUT, {1, 2}), double);          \
+        CHECK_RESULT_TYPE(nanvar(INPUT, {1, 2}), double);           \
+        CHECK_RESULT_TYPE(nanstd(INPUT, {1, 2}), double);           \
+        CHECK_RESULT_TYPE(nanmin(INPUT, {1, 2}), MINMAX_TYPE);      \
+        CHECK_RESULT_TYPE(nanmax(INPUT, {1, 2}), MINMAX_TYPE);
+
+#define CHECK_TEMPLATED_RESULT_TYPE_FOR_ALL(INPUT, TEMPLATE_TYPE, RESULT_TYPE, STD_TYPE, MINMAX_TYPE)  \
```

Comment:
The macro now becomes a bit more bulky but I guess there will be no more nan functions with unique return types.

Reply:
The return type is defined as `decltype(functor_type(T(), E::value_type));`. This way it is consistent with promotion and arithmetic rules of C++ (using `T() + E::value_type()` would be inconsistent with operations which don't promote `short` to `int` when performing a computation on two `short`).

Reply:
Thanks for the explanation! Actually the design makes sense. I will dig into the code to learn more details.

Reply:
You're welcome! Do not hesitate if you have further questions (here or on the [gitter channel](https://gitter.im/QuantStack/Lobby)